internal/wal_show_handler.go

```diff
@@ -0,0 +1,212 @@
+	SwitchPointLsn   uint64          `json:"switchPointLsn"`
+	StartSegment     string          `json:"startSegment"`
+	EndSegment       string          `json:"endSegment"`
+	SegmentsCount    int             `json:"segmentsCount"`
+	MissingSegments  []string        `json:"missingSegments"`
+	Backups          []*BackupDetail `json:"availableBackups,omitempty"`
+	SegmentRangeSize uint64          `json:"segmentRangeSize"`
+	Status           string          `json:"status"`
+}
+
+func NewTimelineInfo(walSegments *WalSegmentsSequence, historyRecords []*TimelineHistoryRecord) (*TimelineInfo, error) {
+	timelineInfo := &TimelineInfo{
+		Id:               walSegments.timelineId,
+		StartSegment:     walSegments.minSegmentNo.getFilename(walSegments.timelineId),
+		EndSegment:       walSegments.maxSegmentNo.getFilename(walSegments.timelineId),
+		SegmentsCount:    len(walSegments.walSegmentNumbers),
+		SegmentRangeSize: uint64(walSegments.maxSegmentNo-walSegments.minSegmentNo) + 1,
+		Status:           TimelineOkStatus,
+	}
+
+	missingSegments, err := walSegments.FindMissingSegments()
+	if err != nil {
+		return nil, err
+	}
+	timelineInfo.MissingSegments = make([]string, 0, len(missingSegments))
+	for _, segment := range missingSegments {
+		timelineInfo.MissingSegments = append(timelineInfo.MissingSegments, segment.GetFileName())
+	}
+
+	if len(timelineInfo.MissingSegments) > 0 {
+		timelineInfo.Status = TimelineLostSegmentStatus
+	}
+
+	// set parent timeline id and timeline switch LSN if have .history record available
+	if len(historyRecords) > 0 {
+		switchHistoryRecord := historyRecords[len(historyRecords)-1]
+		timelineInfo.ParentId = switchHistoryRecord.timeline
+		timelineInfo.SwitchPointLsn = switchHistoryRecord.lsn
+	}
+	return timelineInfo, nil
+}
+
+// WalSegmentsSequence represents some collection of wal segments with the same timeline
+type WalSegmentsSequence struct {
+	timelineId        uint32
+	walSegmentNumbers map[WalSegmentNo]bool
+	minSegmentNo      WalSegmentNo
+	maxSegmentNo      WalSegmentNo
+}
+
+func NewSegmentsSequence(id uint32, segmentNo WalSegmentNo) *WalSegmentsSequence {
+	walSegmentNumbers := make(map[WalSegmentNo]bool)
+	walSegmentNumbers[segmentNo] = true
+
+	return &WalSegmentsSequence{
+		timelineId:        id,
+		walSegmentNumbers: walSegmentNumbers,
+		minSegmentNo:      segmentNo,
+		maxSegmentNo:      segmentNo,
+	}
+}
+
+// AddWalSegmentNo adds the provided segment number to collection
+func (data *WalSegmentsSequence) AddWalSegmentNo(number WalSegmentNo) {
+	data.walSegmentNumbers[number] = true
+	if data.minSegmentNo > number {
+		data.minSegmentNo = number
+	}
+	if data.maxSegmentNo < number {
+		data.maxSegmentNo = number
+	}
+}
+
+// FindMissingSegments finds missing segments in range [minSegmentNo, maxSegmentNo]
+func (data *WalSegmentsSequence) FindMissingSegments() ([]*WalSegmentDescription, error) {
+	startWalSegment := WalSegmentDescription{number: data.maxSegmentNo, timeline: data.timelineId}
```

Comment:
-> start
-> maxSegment
minSegment?

Reply:
the runner is meant to be run in reversed chronological order so we start from the latest